### PR TITLE
feat: add flash plugin

### LINKED ISSUE #1455: Add support for "Message Flashing"
### Discussed in https://github.com/orgs/litestar-org/discussions/1454

<div type='discussions-op-text'>

<sup>Originally posted by **4n1qz5skwv** April  6, 2023</sup>
https://flask.palletsprojects.com/en/2.2.x/patterns/flashing/

```python
flash('Invalid password provided', 'error')
```

How to do this **LiteStar**?</div>

<a href="https://polar.sh/litestar-org/litestar/issues/1455">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/1455/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by euri10**
## Description

- added a flash plugin ala flask, uses the request state

## Closes

Closes #1455, #2560

### CODE REVIEW

File: litestar/contrib/minijinja.py [hunk 1]
```diff
@@ -172,7 +172,7 @@ def register_template_callable(
         Returns:
             None
         """
-        self.engine.add_global(key, pass_state(template_callable))
+        self.engine.add_global(key, pass_state(_transform_state(template_callable)))
```
Comment (euri10) [inline litestar/contrib/minijinja.py:175]: I needed to do this unrelated change in order to be ble to access the state while using this engine

File: litestar/plugins/flash.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        """Register the message callable on the template engine instance."""
+        self.config.template_config.engine_instance.register_template_callable("get_flashes", get_flashes)
+        return app_config
+
+
+def flash(connection: ASGIConnection, message: str, category: str) -> None:
+    """Add a flash message to the request scope."""
+    if "_flash_messages" not in connection.state:
+        connection.state["_flash_messages"] = []
+    connection.state["_flash_messages"].append({"message": message, "category": category})
```
Comment (provinzkraut) [inline litestar/plugins/flash.py:44]: Using `ScopeState` would be appropriate here (in `get_flashes` as well)

**Comment by euri10** [comment]
```diff
diff --git a/docs/examples/templating/template_functions_minijinja.py b/docs/examples/templating/template_functions_minijinja.py
index 0fccdccb3..5655f92e5 100644
--- a/docs/examples/templating/template_functions_minijinja.py
+++ b/docs/examples/templating/template_functions_minijinja.py
@@ -26,4 +26,4 @@ def index() -> Template:
     return Template(template_name="index.html.minijinja")
 
 
-app = Litestar(route_handlers=[index], template_config=template_config)
+app = Litestar(route_handlers=[index], template_config=template_config, debug=True)
diff --git a/litestar/contrib/minijinja.py b/litestar/contrib/minijinja.py
index 6829ff45c..f7775a0c0 100644
--- a/litestar/contrib/minijinja.py
+++ b/litestar/contrib/minijinja.py
@@ -7,7 +7,6 @@ from typing import TYPE_CHECKING, Any, Mapping, Protocol, TypeVar, cast
 from typing_extensions import ParamSpec
 
 from litestar.exceptions import ImproperlyConfiguredException, MissingDependencyException, TemplateNotFoundException
-from litestar.plugins.flash import get_flashes as _get_flashes
 from litestar.template.base import (
     TemplateCallableType,
     TemplateEngineProtocol,
@@ -162,17 +161,24 @@ class MiniJinjaTemplateEngine(TemplateEngineProtocol["MiniJinjaTemplate", StateP
         return MiniJinjaTemplate(self.engine, template_name)
 
     def register_template_callable(
-        self, key: str, template_callable: TemplateCallableType[StateProtocol, P, T]
+        self, key: str, template_callable: TemplateCallableType[StateProtocol, P, T], transform: bool = False
     ) -> None:
         """Register a callable on the template engine.
 
         Args:
             key: The callable key, i.e. the value to use inside the template to call the callable.
             template_callable: A callable to register.
+            transform: If True, the callable will be transformed to receive a ``StateProtocol`` instance as first argument.
 
         Returns:
             None
         """
+
+        def is_decorated(func):
+            return hasattr(func, "__wrapped__") or func.__name__ not in globals()
+
+        if not is_decorated(template_callable) or transform:  # or comes from a plugin...
+            template_callable = _transform_state(template_callable)
         self.engine.add_global(key, pass_state(template_callable))
 
     def render_string(self, template_string: str, context: Mapping[str, Any]) -> str:
@@ -206,9 +212,6 @@ def _minijinja_from_state(func: Callable, state: StateProtocol, *args: Any, **kw
     return cast(str, func(template_context, *args, **kwargs))
 
 
-get_flashes = _transform_state(_get_flashes)
-
-
 def __getattr__(name: str) -> Any:
     if name == "minijinja_from_state":
         warn_deprecation(
diff --git a/litestar/plugins/flash.py b/litestar/plugins/flash.py
index 0ffde4a55..4b34d7897 100644
--- a/litestar/plugins/flash.py
+++ b/litestar/plugins/flash.py
@@ -1,10 +1,10 @@
-import contextlib
 from dataclasses import dataclass
 from enum import Enum
 from typing import Any, Mapping
 
 from litestar.config.app import AppConfig
 from litestar.connection import ASGIConnection
+from litestar.contrib.minijinja import MiniJinjaTemplateEngine
 from litestar.plugins import InitPluginProtocol
 from litestar.template import TemplateConfig
 from litestar.template.base import _get_request_from_context
@@ -35,18 +35,10 @@ class FlashPlugin(InitPluginProtocol):
 
     def on_app_init(self, app_config: AppConfig) -> AppConfig:
         """Register the message callable on the template engine instance."""
-        with contextlib.suppress(ImportError):
-            # minijinja needs a different callable here
-            from litestar.contrib.minijinja import MiniJinjaTemplateEngine
-            from litestar.contrib.minijinja import get_flashes as minijinja_get_flashes
-
-            if isinstance(self.config.template_config.engine_instance, MiniJinjaTemplateEngine):
-                self.config.template_config.engine_instance.register_template_callable(
-                    "get_flashes", minijinja_get_flashes
-                )
-                return app_config
-
-        self.config.template_config.engine_instance.register_template_callable("get_flashes", get_flashes)
+        if isinstance(self.config.template_config.engine_instance, MiniJinjaTemplateEngine):
+            self.config.template_config.engine_instance.register_template_callable("get_flashes", get_flashes, True)
+        else:
+            self.config.template_config.engine_instance.register_template_callable("get_flashes", get_flashes)
         return app_config
```

@provinzkraut this is a small diff between your semi-ugly fix and my semi-ugly one.
Honestly I dont have a real preference, my gut feeling just tells me somehow we should do more in the minijinja implementation to detect if the callable we wanna register is or is not wrapped correctly, or defer that responsability to the caller (but would that be breaking ?)
that's an interesting case because imho the plugin should not have to do that switch dance in the app_init and that should be handled oob in the implementation but I dont see an easy way in minijinja's case to tell hey I've been called by a plugin, pass me the state

**Comment by euri10** [comment]
it would be somewhat better to pass an enum of the possible message categories to the FlashConfig so that the category passed to flash(message, category) belongs to that enum only, but I dont know if this is feasible.
the mypy errors seems unrelated to me

File: litestar/plugins/flash.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    """Flash messages Plugin."""
+
+    def __init__(self, config: FlashConfig):
+        """Initialize the plugin."""
+        self.config = config
+
+    def on_app_init(self, app_config: AppConfig) -> AppConfig:
+        """Register the message callable on the template engine instance."""
+        if isinstance(self.config.template_config.engine_instance, MiniJinjaTemplateEngine):
+            from litestar.contrib.minijinja import _transform_state
```
Comment (JacobCoffee) [inline litestar/plugins/flash.py:30]: is it okay to be calling a private method inside a public one?

**Review by JacobCoffee — APPROVED** [review]
this is awesome.

I have put up https://github.com/euri10/litestar/pull/1 for documentation - i can work on the placeholder data if you dont feel like if it gets merged

File: litestar/plugins/flash.py (hunk 3 above)
Comment (euri10) [reply litestar/plugins/flash.py:30]: honestly I'm not real happy about it, but the minijinja implementation has that particularity of needing that "tweak", I'm really happy if there is another way, @provinzkraut had something similar a few commits above, I find this solution a little bit simpler but have no real preference of one over another, maybe someone would come up with something better

Comment (provinzkraut) [reply litestar/plugins/flash.py:30]: Well this interface isn't great, I just did this as a poc to fix the bug :grimacing:

**Comment by euri10** [comment]
> this is awesome.
> 
> I have put up [euri10#1](https://github.com/euri10/litestar/pull/1) for documentation - i can work on the placeholder data if you dont feel like if it gets merged

sorry I'm slow and didn't understood at first it was on my fork :) need to wake up !

File: litestar/plugins/flash.py (hunk 3 above)
Comment (euri10) [reply litestar/plugins/flash.py:42]: in essence this is kind of minijina's implementation fault (another stuff of mine :-1: ) I have to resort to this, maybe a more elegant solution would be to rethink how minijinja is implemented in order not to have to resort to swith and private import, but I failed at finding a way to be fair

**Comment by euri10** [comment]
working on docs, back to draft mode

**Comment by JacobCoffee** [comment]
Sorry i caught this late, feel free to disregard - i just thought about the API docs and added those but ended up doing more.

not 100% on the validity.

https://github.com/euri10/litestar/pull/2

anyways its working good for me in testing ty for your PR :)

**Review by JacobCoffee — APPROVED** [review]

**Comment by JacobCoffee** [comment]
commit tree looks weird

**Comment by provinzkraut** [comment]
> commit tree looks weird

Needs a rebase

**Comment by euri10** [comment]
honestly I dont get the rebase and stuff, now we have a PR with 127 changes :(
should I just close this and redo from scratch ?

**Comment by euri10** [comment]
are we better now ?

**Comment by provinzkraut** [comment]
> are we better now ?

Looks good! You did it! :)

**Comment by euri10** [comment]
> > are we better now ?
> 
> Looks good! You did it! :)

@Alc-Alc did ;)

**Comment by JacobCoffee** [comment]
Failing types arent even related to PR :(
https://github.com/litestar-org/litestar/actions/runs/8223737329/job/22486749530?pr=3145

**Comment by euri10** [comment]
> Failing types arent even related to PR :( [litestar-org/litestar/actions/runs/8223737329/job/22486749530?pr=3145](https://github.com/litestar-org/litestar/actions/runs/8223737329/job/22486749530?pr=3145)

I had a pdm.lock change, hopefully this will solve it

**Review by JacobCoffee — APPROVED** [review]

**Pull request merged**